pandas/core/indexes/datetimelike.py

```diff
@@ -95,18 +95,29 @@ def _round(self, freq, rounder):
         unit = to_offset(freq).nanos
         # round the local times
         values = _ensure_datetimelike_to_i8(self)
-        if unit < 1000 and unit % 1000 != 0:
+
+        if unit < 1000:
             # for nano rounding, work with the last 6 digits separately
             # due to float precision
             buff = 1000000
             result = (buff * (values // buff) + unit *
-                      (rounder((values % buff) / float(unit))).astype('i8'))
-        elif unit >= 1000 and unit % 1000 != 0:
-            msg = 'Precision will be lost using frequency: {}'
-            warnings.warn(msg.format(freq))
-            result = (unit * rounder(values / float(unit)).astype('i8'))
+                      (rounder((values % buff) * (1 / float(unit))))
+                      .astype('i8'))
         else:
-            result = (unit * rounder(values / float(unit)).astype('i8'))
+            if unit % 1000 != 0:
+                msg = 'Precision will be lost using frequency: {}'
+                warnings.warn(msg.format(freq))
+
+            # GH19206
+            # to deal with round-off when unit is large
+            if unit >= 1e9:
+                divisor = 10 ** int(np.log10(unit / 1e7))
```

Comment:
we should probably pull out all of this rounding logic to a separate function that can be called on a scalar & and array, so we can use it both for Timestamp and DTI (could do it in this PR or as a followup)

Reply:
I think that's a good idea. We can do it here while this is still open.

Reply:
great!